### PR TITLE
[JUJU-4638] Drop charm.URL in flight where possible

### PR DESCRIPTION
**Pull request by jack-w-shaw**
Replace `charm.URL` with `string` in a number of places endemic to the code

Most noteworthy, `st.Charm` now takes a string param instead of a `charm.URL`. This has many ramifications throughout the codebase. However, generalizing this methods signature in this was is a significant step towards dropping series out of charm url entirely. It will now be much easier to pass in something completely different

Also the `Charm` structure in `api/agent/uniter` no longer has a `String` method, instead `URL` returns a stringified url. It doesn't really make sense to me that the string representation of the charm api is the charm URL. use `.URL()` instead

There are a couple of places where we retain charm.URLs, are extra places where we (semi-redundantly) parse strings into charm.URL. This is because the number of functions taking charm.URL is very large, and to keep the size of this PR down, instead of change them all, we parse at the edge 

This PR should result in no changes in behavior

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- ~[ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

All unit test pass 

A number of specific integration test pass:
```
./main.sh -v -c aws -p ec2 deploy test_deploy_charms
./main.sh -v -c aws -p ec2 refresh
```

```
juju bootstrap lxd lxd
juju add-model m
juju deploy ubuntu
juju deploy postgresql
```

### CODE REVIEW

File: apiserver/facades/client/client/status.go [hunk 1]
```diff
@@ -1290,9 +1290,13 @@ func (context *statusContext) processApplication(application *state.Application)
...
-	if latestCharm, ok := context.allAppsUnitsCharmBindings.latestCharms[*applicationCharm.URL().WithRevision(-1)]; ok && latestCharm != nil {
-		if latestCharm.Revision() > applicationCharm.URL().Revision {
-			processedStatus.CanUpgradeTo = latestCharm.String()
+	curl, err := charm.ParseURL(applicationCharm.URL())
+	if err != nil {
+		return params.ApplicationStatus{Err: apiservererrors.ServerError(err)}
+	}
+	if latestCharm, ok := context.allAppsUnitsCharmBindings.latestCharms[*curl.WithRevision(-1)]; ok && latestCharm != nil {
+		if latestCharm.Revision() > curl.Revision {
+			processedStatus.CanUpgradeTo = latestCharm.URL()
```
Comment (hmlanigan) [inline apiserver/facades/client/client/status.go:1299]: This will be interesting for planning the removal of charm url - there are implication for the charmrevisionupdaterworker

File: apiserver/facades/client/machinemanager/upgradeseries.go [hunk 2]
```diff
@@ -464,7 +465,11 @@ func (s stateSeriesValidator) verifySupportedBase(application Application, base
 		return errors.Trace(err)
 	}
 	if len(supportedSeries) == 0 {
-		supportedSeries = append(supportedSeries, ch.URL().Series)
+		curl, err := charm.ParseURL(ch.URL())
+		if err != nil {
+			return errors.Trace(err)
+		}
+		supportedSeries = append(supportedSeries, curl.Series)
 	}
```
Comment (hmlanigan) [inline apiserver/facades/client/machinemanager/upgradeseries.go:473]: same regarding charms with no series. - That should be in a separate commit? Or explained somewhere

File: state/application.go [hunk 3]
```diff
@@ -2214,15 +2210,19 @@ func (a *Application) VerifySupportedBase(base Base, force bool) error {
 		return errors.Trace(err)
 	}
 	if len(supportedSeries) == 0 {
-		supportedSeries = append(supportedSeries, ch.URL().Series)
+		curl, err := charm.ParseURL(ch.URL())
+		if err != nil {
+			return err
+		}
+		supportedSeries = append(supportedSeries, curl.Series)
```
Comment (hmlanigan) [inline state/application.go:2217]: same again

**Review by hmlanigan — CHANGES_REQUESTED** [review]
What have you tested with this change so far? Deploy is just the start.

These are my comments till now - 10 files more to do

File: state/charm.go [hunk 4]
```diff
@@ -627,30 +627,21 @@ func (c *Charm) globalKey() string {
...
-	return c.charmURL
-}
-
 // Revision returns the monotonically increasing charm
 // revision number.
 // Parse the charm's URL on demand, if required.
 func (c *Charm) Revision() int {
 	if c.charmURL == nil {
-		c.charmURL = charm.MustParseURL(c.String())
+		c.charmURL = charm.MustParseURL(c.URL())
```
[the remark this answers is not in the document]
Comment (jack-w-shaw) [reply state/charm.go:644]: Yes very true. Would the answer simply be to store the charm revision in Mongo? We could then mutate, or create a new entry, when a revision is changed (depending how we use the charms)

File: state/state.go [hunk 5]
```diff
@@ -909,12 +909,7 @@ func (st *State) FindEntity(tag names.Tag) (Entity, error) {
 		}
 		return model.Operation(tag.Id())
 	case names.CharmTag:
```
[the remark this answers is not in the document]
Comment (jack-w-shaw) [reply state/state.go:911]: Idk really what this function is used for and where, but my instinct is it would be better to keep it, so that in the future if we do want to use it with a charm tag it still works

File: apiserver/facades/client/application/deployrepository.go [hunk 6]
```diff
@@ -806,36 +815,36 @@ func (v *deployFromRepositoryValidator) getCharm(arg params.DeployFromRepository
...
-		return resolvedData.URL, resolvedOrigin, deployedCharm, nil
+		return resolvedData.URL.String(), resolvedOrigin, deployedCharm, nil
 	}
 
 	// This charm needs to be downloaded, remove the ID and Hash to
 	// allow it to happen.
 	resolvedOrigin.ID = ""
 	resolvedOrigin.Hash = ""
-	return resolvedData.URL, resolvedOrigin, resolvedCharm, nil
+	return resolvedData.URL.String(), resolvedOrigin, resolvedCharm, nil
```
[the remark this answers is not in the document]
Comment (jack-w-shaw) [reply apiserver/facades/client/application/deployrepository.go:847]: Because ideally in the next step (or one of the next steps) I wouldn't need to parse the url to get the revision, or to construct a charm.ID

However, as with an above comment, this can be done when this is implemented

File: apiserver/facades/client/bundle/bundle.go [hunk 7]
```diff
@@ -542,7 +542,7 @@ func (b *BundleAPI) bundleDataApplications(
 			// from the charm config metadata.
 			cfgInfo, ok := charmConfigCache[charmURL]
 			if !ok {
-				ch, err := backend.Charm(curl)
+				ch, err := backend.Charm(application.CharmURL())
```
[the remark this answers is not in the document]
Comment (jack-w-shaw) [reply apiserver/facades/client/bundle/bundle.go:545]: I don't think application.CharmURL does call into state? I think this could be:
https://github.com/juju/juju/blob/3.3/core/cache/application.go#L69-L71

However, this call here is on an interface, so i think you're right in principle, it probably is good practise to store the result as we don't know how costly this is

File: apiserver/facades/client/application/updateseries.go [hunk 8]
```diff
@@ -141,7 +142,11 @@ func (s stateSeriesValidator) ValidateApplication(application Application, base
 		return errors.Trace(err)
 	}
 	if len(supportedSeries) == 0 {
-		supportedSeries = append(supportedSeries, ch.URL().Series)
+		curl, err := charm.ParseURL(ch.URL())
+		if err != nil {
+			return errors.Trace(err)
+		}
+		supportedSeries = append(supportedSeries, curl.Series)
 	}
```
[the remark this answers is not in the document]
Comment (jack-w-shaw) [reply apiserver/facades/client/application/updateseries.go:150]: I'll remove this in a separate commit

File: apiserver/facades/client/machinemanager/upgradeseries.go (hunk 2 above)
Comment (jack-w-shaw) [reply apiserver/facades/client/machinemanager/upgradeseries.go:473]: I'll remove this in a separate commit

File: state/application.go (hunk 3 above)
Comment (jack-w-shaw) [reply state/application.go:2217]: I'll remove this in a separate commit

**Review by hmlanigan — APPROVED** [review]
